streams/src/test/java/org/apache/kafka/streams/integration/KStreamRestorationIntegrationTest.java

```diff
@@ -0,0 +1,122 @@
+    private final MockTime mockTime = CLUSTER.time;
+
+    @Before
+    public void setUp() throws Exception {
+        final Properties props = new Properties();
+
+        streamsConfiguration = StreamsTestUtils.getStreamsConfig(
+                APPLICATION_ID,
+                CLUSTER.bootstrapServers(),
+                Serdes.Integer().getClass().getName(),
+                Serdes.ByteArray().getClass().getName(),
+                props);
+
+        CLUSTER.createTopics(INPUT_TOPIC);
+        CLUSTER.createTopics(OUTPUT_TOPIC);
+
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+    }
+
+    @Test
+    public void shouldRestoreNullRecord() throws InterruptedException, ExecutionException {
+        builder.table(INPUT_TOPIC, Materialized.<Integer, byte[]>as(
+                Stores.persistentTimestampedKeyValueStore(STATE_STORE_NAME))
+                .withKeySerde(Serdes.Integer())
+                .withValueSerde(Serdes.ByteArray())
+                .withCachingDisabled()).toStream().to(OUTPUT_TOPIC);
+
+        final Properties producerConfig = TestUtils.producerConfig(
+                CLUSTER.bootstrapServers(), IntegerSerializer.class, ByteArraySerializer.class);
+
+        final List<KeyValue<Integer, byte[]>> initialKeyValues = Arrays.asList(
+                KeyValue.pair(3, new byte[]{3}), KeyValue.pair(3, null), KeyValue.pair(1, new byte[]{1}));
+
+        IntegrationTestUtils.produceKeyValuesSynchronously(
+                INPUT_TOPIC, initialKeyValues, producerConfig, mockTime);
+
+        KafkaStreams streams = new KafkaStreams(builder.build(streamsConfiguration), streamsConfiguration);
+        streams.start();
+
+        final Properties consumerConfig = TestUtils.consumerConfig(
+                CLUSTER.bootstrapServers(), IntegerDeserializer.class, ByteArrayDeserializer.class);
+        // We couldn't use a #waitUntilFinalKeyValueRecordsReceived here because the byte array comparison is not triggered correctly.
+        final List<KeyValue<Integer, byte[]>> outputs = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(
+                consumerConfig, OUTPUT_TOPIC, 3);
+        for (int i = 0; i < outputs.size(); i++) {
+            assertEquals(initialKeyValues.get(i).key, outputs.get(i).key);
+            assertArrayEquals(initialKeyValues.get(i).value, outputs.get(i).value);
+        }
+
+        streams.close();
+        streams.cleanUp();
+
+        // Restart the stream instance. There should not be exception handling the null value within changelog topic.
+        IntegrationTestUtils.produceKeyValuesSynchronously(
+                INPUT_TOPIC, Collections.singletonList(KeyValue.pair(2, new byte[3])), producerConfig, mockTime);
+        streams = new KafkaStreams(builder.build(streamsConfiguration), streamsConfiguration);
+        streams.start();
+        IntegrationTestUtils.waitUntilMinRecordsReceived(consumerConfig, OUTPUT_TOPIC, 1);
+        streams.close();
```

Comment:
I think we should check that the received record is the new kv-pair with key `2`.